Spreadsheets can be loaded from file-like objects and random functions

doconix opened the pull request:
1. While openpyxl allows loading spreadsheets from both file paths (str) and file-like objects, koala only supported file paths. The Spreadsheet constructor now checks the `file` parameter for a `.read` attribute, and if it has this attribute, assumes it is an open stream to a file rather than the file path string.
2. Added Excel's two random (RAND and RANDBETWEEN) functions to the supported functions. Since python's random library is not thread-safe (threads would share the same seed), the code creates the random.Random() object directly so each call to the function has its own seed and is thread safe.

I also added two unit tests to test both ways of loading spreadsheets.

doconix committed:
Spreadsheets can be loaded from file-like objects and two new functions: RAND and RANDBETWEEN

1. While openpyxl allows loading spreadsheets from both file paths (str) and file-like objects, koala only supported file paths. The Spreadsheet constructor now checks the `file` parameter for a `.read` attribute, and if it has this attribute, assumes it is an open stream to a file rather than the file path string.
2. Added Excel's two random functions to the supported functions. Since python's random library is not thread-safe (threads would share the same seed), the code creates the random.Random() object directly so each call to the function has its own seed and is thread safe.

I also added two unit tests to test both ways of loading spreadsheets.

danielsjf commented:
@doconix, great work, thanks for these changes. Could you also add a few tests for the rand functions? The file reading is fine.

doconix commented:
Yes will do

doconix commented:
OK got them in. I can't think of any "better" way to test these (tests are pretty simple). Let me know if you can think of further ways and I can implement.

vallettea commented:
looks good, big thanks.
last thing can you just increase the version number in package.json so i can push it to pipy

doconix commented:
I updated the version number to 0.32 in `setup.py`. (I think that's what you meant... ;)

I also squashed the three commits into a single commit that goes with this PR so the git log is cleaner.

Let me know if you see anything else!

doconix committed:
Small refactor of `Spreadsheet.from_dict` to make it faster.

After profiling the code to see why `from_dict` was slow, it was obvious that
the inner function `find_cell` was the culprit. This refactor removes the
inner function and instead uses a temporary dictionary to make things fast.

The two profile outputs below show load time going from 2.42 -> .23 seconds.

===========================

Code used to test (this uses a fairly large, complex spreadsheet):
    from koala import Spreadsheet
    import cProfile, pstats
    pr = cProfile.Profile()
    file_name = '/Users/conan/Downloads/logicandreference-a9-Conan_Albrecht-1.xlsm'
    sp1 = Spreadsheet(file_name)
    data = sp1.asdict()
    pr.enable()
    for i in range(5):
        sp2 = Spreadsheet.from_dict(data)
    pr.disable()
    ps = pstats.Stats(pr).sort_stats('time')
    ps.print_stats(50)

Before the change:
         7158776 function calls (7158276 primitive calls) in 2.418 seconds

   Ordered by: internal time
   List reduced from 83 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     5520    1.408    0.000    2.202    0.000 /Users/conan/Documents/data/programming/koala/koala/Spreadsheet.py:1073(find_cell)
  6889145    0.797    0.000    0.797    0.000 /Users/conan/Documents/data/programming/koala/koala/Cell.py:172(address)
     2135    0.050    0.000    0.050    0.000 {built-in method builtins.compile}
    10975    0.047    0.000    0.067    0.000 /Users/conan/Documents/data/programming/koala/koala/Cell.py:21(__init__)

After the change:
         297041 function calls (296541 primitive calls) in 0.230 seconds

   Ordered by: internal time
   List reduced from 83 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    10975    0.063    0.000    0.083    0.000 /Users/conan/Documents/data/programming/koala/koala/Cell.py:21(__init__)
     2135    0.041    0.000    0.041    0.000 {built-in method builtins.compile}
        5    0.024    0.005    0.053    0.011 /Users/conan/.pyenv/versions/me3.6/lib/python3.6/site-packages/networkx/readwrite/json_graph/node_link.py:104(node_link_graph)

danielsjf commented:
@doconix, it seems that similar changes ended up in the two pull requests. Could you disentangle them and clean the pulls? This now also has a speed refactor.

danielsjf commented:
@doconix, after merging the other pull request, the issue is clearer.

Also, could you change the version to 0.0.34? Somehow, the version on Pypi is already 0.0.33 (granted, there seems to be a misalignment with the version on Github so let's fix that).

doconix committed:
fixed merge conflicts

doconix committed:
version bump to 0.0.34

doconix commented:
I somehow included the speedup refactor code in the previous PR. Apologies for that mixup. It was supposed to be a different branch/PR.

I've updated the remaining PR to have version 0.0.34, and the diff seems to be clear now (as you mention). 

Thanks for looking at these PRs -- my company creates college-level online textbooks, and we are using koala to help grade embedded questions. Being able to convert Excel over to pure python is fantastic. These PRs are because we needed a few extra things to integrate it right, such as loading from S3 file streams rather than the server filesystem.

Let me know if you see anything else. Conan

danielsjf commented:
@doconix, great to see all these uses and thanks a lot for your contribution.

@vallettea, could you update this version on Pypi?

vallettea commented:
sorry for the delay, i updated on pipy
